implementation/src/main/java/io/smallrye/jwt/auth/cdi/RawClaimTypeProducer.java

```diff
@@ -121,18 +121,14 @@ Double getClaimAsDouble(InjectionPoint ip) {
     }
 
     /**
-     * Produces a *raw* Optional value. The type parameter is intentionally
-     * left unspecified on this producer to avoid ambiguous type resolution
-     * with other producers of Optional values where the type parameter is
-     * specified.
+     * Produces a *raw* Optional value.
      *
      * @param ip reference to the injection point
      * @return an optional claim value
      */
     @Produces
     @Claim("")
-    @SuppressWarnings("rawtypes")
-    public Optional getOptionalValue(InjectionPoint ip) {
+    public <T> Optional<T> getOptionalValue(InjectionPoint ip) {
```

Comment:
Hi Darran @darranl as far as I recall that fails in Thorntail, I've tried it before, can conform tomorrow. Perhaps we can reuse the Quarkus idea somehow instead if it proves a problem...

Reply:
The Quarkus approach seems to rely on an io.quarkus.arc.BeanCreator so that seems Quarkus specific, unless there is a pure CDI equivalent we can use?

I do have a project I could add my own producer to but that is more of a last resort, ideally I think SmallRye JWT should contain the producer and the current definition is not compatible with the CDI specification - however maybe there is an alternative.

For the various feedback I have received on the PRs is feels like maybe some central customisation within the CDI extension could help to provide "tweaks" to the capabilities of SmallRye JWT.

We could either make use of an abstract Extension or add some additional methods to the existing extension but then a child extension could override specific behaviour.  As an example we could have a method: -

protected boolean useUntypedOptionalClaims() {
    return true;
}

Where this returns true the extension can register a producer of the original format and if false register a producer with this new format.

This is a bit of a tangent for this PR but there are other methods that could be useful to override such as isHttpAuthMechanismEnabled() and isEESecurityAvailable() - in terms of the WildFly integration I would hard code both of these to true but by default we would leave the existing behaviour.

Going back to my earlier PR regarding the static vs @ApplicationScoped bean, extension overrides may be a way to better handle that.

Sorry that did diverge quite a bit ;-)

Reply:
@darranl What is interesting it is actually working in Thorntail now :-), @Ladicek updated it to WildFly 18 and I guess it has made all the difference. Checking Quarkus now...

Re your PR #147, yes, lets discuss the other options there (or at the related issue). I was somewhat concerned (without exactly understanding why :-) ) about introducing a new `entity` into the API, such as `JwtParser` - may be because now there would be two ways for the consumers to bootstrap `smallrye-jwt`, via the factory or via this new injected parser. (FYI, in Quarkus, at some point there was some duplication, no factories were used and the code was paste/copied so I move it all into `DefaultJwtParser`). #147 still looks good to me though...

Reply:
+1 on continuing on #147 - if we can find a solution for this one I think #147 is my final issue to have a clean TCK run with SmallRye JWT integrated directly into WildFly.

Reply:
Well, Quarkus does not like it. Let me ask @mkouba

Reply:
With Quarkus I wonder if it conflicts with their workaround?

Reply:
I'm sorry what's the question?

Reply:
Hi Martin @mkouba , was just about to ping you on Zulip dev :-), so with Darran's fix I'm seeing:
```
Suppressed: javax.enterprise.inject.AmbiguousResolutionException: Ambiguous dependencies for type java.util.Optional<javax.json.JsonString> and qualifiers [@Claim(standard = preferred_username)]
	- java member: io.quarkus.jwt.test.DefaultScopedEndpoint#currentUsername
	- declared on CLASS bean [types=[io.quarkus.jwt.test.DefaultScopedEndpoint], qualifiers=[@Default, @Any], target=io.quarkus.jwt.test.DefaultScopedEndpoint]
	- available beans:
		- PRODUCER METHOD bean [types=[java.util.Optional<javax.json.JsonString>, java.lang.Object], qualifiers=[@Any, @Claim(value = "")], target=java.util.Optional<javax.json.JsonString> getOptionalJsonString(javax.enterprise.inject.spi.InjectionPoint ip), declaringBean=io.smallrye.jwt.auth.cdi.JsonValueProducer]
		- PRODUCER METHOD bean [types=[java.util.Optional<T>, java.lang.Object], qualifiers=[@Any, @Claim(value = "")], target=java.util.Optional<T> getOptionalValue(javax.enterprise.inject.spi.InjectionPoint ip), declaringBean=io.smallrye.jwt.auth.cdi.RawClaimTypeProducer]
		at io.quarkus.arc.processor.Beans.resolveInjectionPoint(Beans.java:487)
		at io.quarkus.arc.processor.BeanInfo.init(BeanInfo.java:416)
		at io.quarkus.arc.processor.BeanDeployment.init(BeanDeployment.java:215)
```
which is probably correct if Quarkus also registers an explicit producer for JSON..., may be somewhere here:

https://github.com/quarkusio/quarkus/blob/master/extensions/smallrye-jwt/deployment/src/main/java/io/quarkus/smallrye/jwt/deployment/SmallRyeJwtProcessor.java#L123

FYI, @darranl would like to have something like this supported:
```
// not only JSON String:
@Inject
@Claim(standard = Claims.birthdate)
Optional<JsonString> birthdate;

// or, just:
@Inject
@Claim(standard = Claims.birthdate)
Optional<String> birthdate;

```

Reply:
@mkouba I wonder if we can move some CDI code out of Quarkus to smallrye-jwt to avoid the conflicts. FYI, Darran's fix is OK in the latest Thorntail

Reply:
@mkouba actually Thorntail does not have a single test checking `JsonString`, let me check that too

Reply:
Yes, it's correct because according to the typesafe resolution rules a bean with type `Optional<T>` is assignable to the `Optional<JsonString>` required type:
```
A parameterized bean type is considered assignable to a parameterized required type if they have identical raw type and for each parameter:
* the required type parameter is an actual type, the bean type parameter is a type variable and the actual type is assignable to the upper bound, if any, of the type variable,
```
In this case, `JsonString` is assignable to `java.lang.Object`.

For a raw type (`Optional`) a different rule applies:
```
A raw bean type is considered assignable to a parameterized required type if the raw types are identical and all type parameters of the required type are either unbounded type variables or java.lang.Object.
```

Reply:
@manovotn Pls do a quick sanity check ^ ;-)

Reply:
@sberyozkin it should NOT work in TT!

Reply:
@mkouba Hi Martin, so this is why I'm thinking aloud that we should have a piece of CDI code in `smallrye-jwt` which can be reused in Quarkus/WildFly/Thorntail, in Quarkus this would be shared code would be used one way, in WildFly - as part of some CDI extension, etc.

Reply:
The problem is that we can't reuse the CDI extensions in Quarkus. Of course, producers and similar stuff could be reused.

Reply:
Just to clarify, this is not that I would like Optional<String> claim injection supported,  I was going to try and adapt the quickstart to avoid that if it is not presently possible,

However the MicroProfile JWT TCK specifically tests typed Optional claim injection and as it is defined in the specification I don't think we can challenge the TCK.

Reply:
> In this case, JsonString is assignable to java.lang.Object.

Yes, that's exactly the case.

> For a raw type (Optional) a different rule applies:

Yeah, basically `Optional` can be assigned to something like `Optional<T>` and `Optional<Object>`

Reply:
@mkouba yeah, we'd only have something similar to [ProviderExtensionSupport](https://github.com/smallrye/smallrye-jwt/blob/master/implementation/src/main/java/io/smallrye/jwt/auth/cdi/ProviderExtensionSupport.java). In Thorntail the CDI extension extends it, but in Quarkus it can be used as a helper utility code. So we'd have some shared code which would be reused somehow differently in different projects.

Reply:
@mkouba @manovotn Hi Martin, Matej sure, this is now clear :-). So is there a chance some code which makes it work in Quarkus can be moved to `smallrye-jwt` ? @darranl then can update the PR accordingly.
Martin, yes, it does not work in Thorntail with `<T>`, you are right :-)

Reply:
Which code are we saying can be moved from Quarkus?  The code I have seen is making use of Quarkus APIs to add this support.

Reply:
@darranl I'm not sure, but yes, there is Quarkus API there but it seems some of that code is not tied to Quarkus and can be moved to a `smallrye-jwt` utility class, for it be reused from Quarkus APi or CDI extensions in WildFly/Thorntail. Can you something that might be moved from  [here](https://github.com/quarkusio/quarkus/blob/master/extensions/smallrye-jwt/deployment/src/main/java/io/quarkus/smallrye/jwt/deployment/SmallRyeJwtProcessor.java#L123) and/or [here](https://github.com/quarkusio/quarkus/blob/master/extensions/smallrye-jwt/runtime/src/main/java/io/quarkus/smallrye/jwt/runtime/auth/RawOptionalClaimCreator.java#L17) ?

Reply:
It sounds to me the main aspect is ensuring Optional, Optional<?>, Optional<String + other types>, Optional<JsonString>.

If Quarkus is reporting a "Ambiguous dependencies" error it sounds like the duplicate registration would then need to be removed from within Quarkus.

Reply:
Looks like I can reproduce without Quarkus if I add all the Optional types to my Quickstart.

Reply:
Outside of Optional<JsonValue> there are only four types required to be supported by the spec String, Set<String>, Long, and Boolean - rather than a generic producer I suspect easier to just add each individual one.